# Segmenting and Clustering Neighborhoods in Toronto

## Part I

In [1]:
!pip install bs4

In [104]:
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup # this module helps in web scrapping.
import requests  # this module helps us to download a web page

In [105]:
url = 'https://en.wikipedia.org/w/index.php?title=List_of_postal_codes_of_Canada:_M&direction=prev&oldid=926287641'

In [106]:
data  = requests.get(url).text

In [107]:
soup = BeautifulSoup(data,"html5lib")
#find all html tables in the web page
tables = soup.find_all('table') # in html table is represented by the tag <table>
# we can see how many tables were found by checking the length of the tables list
len(tables)

5

In [108]:
for index,table in enumerate(tables):
    if ("Toronto" in str(table)):
        table_index = index
print(table_index)

0


In [109]:
print(tables[table_index].prettify())

<table class="wikitable sortable">
 <tbody>
  <tr>
   <th>
    Postcode
   </th>
   <th>
    Borough
   </th>
   <th>
    Neighbourhood
   </th>
  </tr>
  <tr>
   <td>
    M1A
   </td>
   <td>
    Not assigned
   </td>
   <td>
    Not assigned
   </td>
  </tr>
  <tr>
   <td>
    M2A
   </td>
   <td>
    Not assigned
   </td>
   <td>
    Not assigned
   </td>
  </tr>
  <tr>
   <td>
    M3A
   </td>
   <td>
    <a href="/wiki/North_York" title="North York">
     North York
    </a>
   </td>
   <td>
    <a href="/wiki/Parkwoods" title="Parkwoods">
     Parkwoods
    </a>
   </td>
  </tr>
  <tr>
   <td>
    M4A
   </td>
   <td>
    <a href="/wiki/North_York" title="North York">
     North York
    </a>
   </td>
   <td>
    <a href="/wiki/Victoria_Village" title="Victoria Village">
     Victoria Village
    </a>
   </td>
  </tr>
  <tr>
   <td>
    M5A
   </td>
   <td>
    <a href="/wiki/Downtown_Toronto" title="Downtown Toronto">
     Downtown Toronto
    </a>
   </td>
   <td>
    <a class=

In [110]:
neigh = pd.DataFrame(columns=["Postal Code", "Borough", "Neighbourhood"])

for row in tables[table_index].tbody.find_all("tr"):
    col = row.find_all("td")
    if (col != []):
        PostalCode = col[0].text
        Borough = col[1].text
        Neighbourhood = col[2].text.strip()
        neigh = neigh.append({"Postal Code":PostalCode, "Borough":Borough, "Neighbourhood":Neighbourhood}, ignore_index=True)

neigh

Postal Code           Borough          Neighbourhood
0           M1A      Not assigned           Not assigned
1           M2A      Not assigned           Not assigned
2           M3A        North York              Parkwoods
3           M4A        North York       Victoria Village
4           M5A  Downtown Toronto           Harbourfront
..          ...               ...                    ...
283         M8Z         Etobicoke              Mimico NW
284         M8Z         Etobicoke     The Queensway West
285         M8Z         Etobicoke  Royal York South West
286         M8Z         Etobicoke         South of Bloor
287         M9Z      Not assigned           Not assigned

[288 rows x 3 columns]

In [111]:
df = neigh.groupby(['Postal Code','Borough'])['Neighbourhood'].apply(','.join).to_frame().reset_index()
df

Postal Code       Borough  \
0           M1A  Not assigned   
1           M1B   Scarborough   
2           M1C   Scarborough   
3           M1E   Scarborough   
4           M1G   Scarborough   
..          ...           ...   
175         M9V     Etobicoke   
176         M9W     Etobicoke   
177         M9X  Not assigned   
178         M9Y  Not assigned   
179         M9Z  Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                        Rouge,Malvern  
2                 Highland Creek,Rouge Hill,Port Union  
3                      Guildwood,Morningside,West Hill  
4                                               Woburn  
..                                                 ...  
175  Albion Gardens,Beaumond Heights,Humbergate,Jam...  
176                                          Northwest  
177                                       Not assigned  
178                                       Not assigned  
179                                       Not assigned  

[180 rows x 3 columns]

In [112]:
df = df[df.Borough != 'Not assigned'].reset_index(drop=True)
df.head(20)

Postal Code      Borough                                      Neighbourhood
0          M1B  Scarborough                                      Rouge,Malvern
1          M1C  Scarborough               Highland Creek,Rouge Hill,Port Union
2          M1E  Scarborough                    Guildwood,Morningside,West Hill
3          M1G  Scarborough                                             Woburn
4          M1H  Scarborough                                          Cedarbrae
5          M1J  Scarborough                                Scarborough Village
6          M1K  Scarborough          East Birchmount Park,Ionview,Kennedy Park
7          M1L  Scarborough                      Clairlea,Golden Mile,Oakridge
8          M1M  Scarborough      Cliffcrest,Cliffside,Scarborough Village West
9          M1N  Scarborough                         Birch Cliff,Cliffside West
10         M1P  Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...
11         M1R  Scarborough                                   Maryvale,Wexford
12         M1S  Scarborough                                          Agincourt
13         M1T  Scarborough              Clarks Corners,Sullivan,Tam O'Shanter
14         M1V  Scarborough  Agincourt North,L'Amoreaux East,Milliken,Steel...
15         M1W  Scarborough                                    L'Amoreaux West
16         M1X  Scarborough                                        Upper Rouge
17         M2H   North York                                  Hillcrest Village
18         M2J   North York                         Fairview,Henry Farm,Oriole
19         M2K   North York                                    Bayview Village

In [113]:
df.Borough.unique()

array(['Scarborough', 'North York', 'East York', 'East Toronto',
       'Central Toronto', 'Downtown Toronto', 'York', 'West Toronto',
       "Queen's Park", 'Mississauga', 'Etobicoke'], dtype=object)

In [114]:
df.shape

(103, 3)

## Part II

In [115]:
geo = pd.read_csv("Geospatial_Coordinates.csv") 
geo

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [116]:
geo2 = geo.drop('Postal Code',axis=1)
geo2

Latitude  Longitude
0    43.806686 -79.194353
1    43.784535 -79.160497
2    43.763573 -79.188711
3    43.770992 -79.216917
4    43.773136 -79.239476
..         ...        ...
98   43.706876 -79.518188
99   43.696319 -79.532242
100  43.688905 -79.554724
101  43.739416 -79.588437
102  43.706748 -79.594054

[103 rows x 2 columns]

In [117]:
neighbourhoods = df.merge(geo, how="inner", on="Postal Code")
neighbourhoods

Postal Code      Borough  \
0           M1B  Scarborough   
1           M1C  Scarborough   
2           M1E  Scarborough   
3           M1G  Scarborough   
4           M1H  Scarborough   
..          ...          ...   
98          M9N         York   
99          M9P    Etobicoke   
100         M9R    Etobicoke   
101         M9V    Etobicoke   
102         M9W    Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                        Rouge,Malvern  43.806686 -79.194353  
1                 Highland Creek,Rouge Hill,Port Union  43.784535 -79.160497  
2                      Guildwood,Morningside,West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
..                                                 ...        ...        ...  
98                                              Weston  43.706876 -79.518188  
99                                           Westmount  43.696319 -79.532242  
100  Kingsview Village,Martin Grove Gardens,Richvie...  43.688905 -79.554724  
101  Albion Gardens,Beaumond Heights,Humbergate,Jam...  43.739416 -79.588437  
102                                          Northwest  43.706748 -79.594054  

[103 rows x 5 columns]

## Parte III

In [118]:
!pip install geopy

In [119]:
!pip install folium

In [145]:
from geopy.geocoders import Nominatim 
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import folium

In [146]:
print('The dataframe has {} boroughs and {} neighbourhoods.'.format(
        len(neighbourhoods['Borough'].unique()),
        neighbourhoods.shape[0]
    )
)

The dataframe has 11 boroughs and 103 neighbourhoods.


In [147]:
address = 'Toronto Canada'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


In [148]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighbourhood in zip(neighbourhoods['Latitude'], neighbourhoods['Longitude'], neighbourhoods['Borough'], neighbourhoods['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

In [149]:
dt_data = neighbourhoods[neighbourhoods['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
dt_data

Postal Code           Borough  \
0          M4W  Downtown Toronto   
1          M4X  Downtown Toronto   
2          M4Y  Downtown Toronto   
3          M5A  Downtown Toronto   
4          M5B  Downtown Toronto   
5          M5C  Downtown Toronto   
6          M5E  Downtown Toronto   
7          M5G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M5J  Downtown Toronto   
10         M5K  Downtown Toronto   
11         M5L  Downtown Toronto   
12         M5S  Downtown Toronto   
13         M5T  Downtown Toronto   
14         M5V  Downtown Toronto   
15         M5W  Downtown Toronto   
16         M5X  Downtown Toronto   
17         M6G  Downtown Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                                            Rosedale  43.679563 -79.377529  
1                          Cabbagetown,St. James Town  43.667967 -79.367675  
2                                Church and Wellesley  43.665860 -79.383160  
3                            Harbourfront,Regent Park  43.654260 -79.360636  
4                             Ryerson,Garden District  43.657162 -79.378937  
5                                      St. James Town  43.651494 -79.375418  
6                                         Berczy Park  43.644771 -79.373306  
7                                  Central Bay Street  43.657952 -79.387383  
8                              Adelaide,King,Richmond  43.650571 -79.384568  
9     Harbourfront East,Toronto Islands,Union Station  43.640816 -79.381752  
10            Design Exchange,Toronto Dominion Centre  43.647177 -79.381576  
11                      Commerce Court,Victoria Hotel  43.648198 -79.379817  
12                      Harbord,University of Toronto  43.662696 -79.400049  
13            Chinatown,Grange Park,Kensington Market  43.653206 -79.400049  
14  CN Tower,Bathurst Quay,Island airport,Harbourf...  43.628947 -79.394420  
15                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
16              First Canadian Place,Underground city  43.648429 -79.382280  
17                                           Christie  43.669542 -79.422564

In [150]:
address = 'Downtown Toronto'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown are 43.6541737, -79.38081162653639.


In [151]:
# create map of Manhattan using latitude and longitude values
map_dt = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(dt_data['Latitude'], dt_data['Longitude'],dt_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_dt)  
    
map_dt

In [152]:
CLIENT_ID = 'QI3MZSSRBG3IQMMRZYVBBQSLS1IIG4VYWX1BOGMKEMPZVPEM' # your Foursquare ID
CLIENT_SECRET = '05S1AQFGOYWX5CXFAMRYQ0TEBWLUS2MS2CXRDEBKARNI40CZ' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: QI3MZSSRBG3IQMMRZYVBBQSLS1IIG4VYWX1BOGMKEMPZVPEM
CLIENT_SECRET:05S1AQFGOYWX5CXFAMRYQ0TEBWLUS2MS2CXRDEBKARNI40CZ


In [153]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [154]:
dt_venues = getNearbyVenues(names=dt_data['Neighbourhood'],
                                   latitudes=dt_data['Latitude'],
                                   longitudes=dt_data['Longitude']
                                  )

Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront,Regent Park
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie


In [155]:
print(dt_venues.shape)
dt_venues.head()

(1205, 7)


Neighbourhood  Neighbourhood Latitude  \
0                    Rosedale               43.679563   
1                    Rosedale               43.679563   
2                    Rosedale               43.679563   
3                    Rosedale               43.679563   
4  Cabbagetown,St. James Town               43.667967   

   Neighbourhood Longitude                 Venue  Venue Latitude  \
0               -79.377529         Rosedale Park       43.682328   
1               -79.377529          Whitney Park       43.682036   
2               -79.377529  Alex Murray Parkette       43.678300   
3               -79.377529        Milkman's Lane       43.676352   
4               -79.367675           Cranberries       43.667843   

   Venue Longitude Venue Category  
0       -79.378934     Playground  
1       -79.373788           Park  
2       -79.382773           Park  
3       -79.373842          Trail  
4       -79.369407          Diner

In [156]:
dt_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Adelaide,King,Richmond                                                  96   
Berczy Park                                                             59   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      16   
Cabbagetown,St. James Town                                              48   
Central Bay Street                                                      63   
Chinatown,Grange Park,Kensington Market                                 63   
Christie                                                                16   
Church and Wellesley                                                    78   
Commerce Court,Victoria Hotel                                          100   
Design Exchange,Toronto Dominion Centre                                100   
First Canadian Place,Underground city                                  100   
Harbord,University of Toronto                                           32   
Harbourfront East,Toronto Islands,Union Station                        100   
Harbourfront,Regent Park                                                47   
Rosedale                                                                 4   
Ryerson,Garden District                                                100   
St. James Town                                                          83   
Stn A PO Boxes 25 The Esplanade                                        100   

                                                    Neighbourhood Longitude  \
Neighbourhood                                                                 
Adelaide,King,Richmond                                                   96   
Berczy Park                                                              59   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                       16   
Cabbagetown,St. James Town                                               48   
Central Bay Street                                                       63   
Chinatown,Grange Park,Kensington Market                                  63   
Christie                                                                 16   
Church and Wellesley                                                     78   
Commerce Court,Victoria Hotel                                           100   
Design Exchange,Toronto Dominion Centre                                 100   
First Canadian Place,Underground city                                   100   
Harbord,University of Toronto                                            32   
Harbourfront East,Toronto Islands,Union Station                         100   
Harbourfront,Regent Park                                                 47   
Rosedale                                                                  4   
Ryerson,Garden District                                                 100   
St. James Town                                                           83   
Stn A PO Boxes 25 The Esplanade                                         100   

                                                    Venue  Venue Latitude  \
Neighbourhood                                                               
Adelaide,King,Richmond                                 96              96   
Berczy Park                                            59              59   
CN Tower,Bathurst Quay,Island airport,Harbourfr...     16              16   
Cabbagetown,St. James Town                             48              48   
Central Bay Street                                     63              63   
Chinatown,Grange Park,Kensington Market                63              63   
Christie                                               16              16   
Church and Wellesley                                   78              78   
Commerce Court,Victoria Hotel                         100             100   
Design Exchange,Toronto Dominion Centre               100      

In [157]:
print('There are {} uniques categories.'.format(len(dt_venues['Venue Category'].unique())))

There are 206 uniques categories.


In [168]:
# one hot encoding
dt_onehot = pd.get_dummies(dt_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
dt_onehot['Neighbourhood'] = dt_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [dt_onehot.columns[-1]] + list(dt_onehot.columns[:-1])
dt_onehot = dt_onehot[fixed_columns]

dt_onehot.head()

Neighbourhood  Adult Boutique  Airport  Airport Food Court  \
0                    Rosedale               0        0                   0   
1                    Rosedale               0        0                   0   
2                    Rosedale               0        0                   0   
3                    Rosedale               0        0                   0   
4  Cabbagetown,St. James Town               0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Thai Restaurant  Theater  \
0                    0             0  ...                0        0   
1                    0             0  ...                0        0   
2                    0             0  ...                0        0   
3                    0             0  ...                0        0   
4                    0             0  ...                0        0   

   Theme Restaurant  Trail  Train Station  Vegetarian / Vegan Restaurant  \
0                 0      0              0                              0   
1                 0      0              0                              0   
2                 0      0              0                              0   
3                 0      1              0                              0   
4                 0      0              0                              0   

   Video Game Store  Vietnamese Restaurant  Wine Bar  Yoga Studio  
0                 0                      0         0            0  
1                 0                      0         0            0  
2                 0                      0         0            0  
3                 0                      0         0            0  
4                 0                      0         0            0  

[5 rows x 207 columns]

In [169]:
dt_onehot.shape

(1205, 207)

In [203]:
dt_grouped = dt_onehot.groupby('Neighbourhood').mean().reset_index()
dt_grouped.head()

Neighbourhood  Adult Boutique  Airport  \
0                             Adelaide,King,Richmond             0.0   0.0000   
1                                        Berczy Park             0.0   0.0000   
2  CN Tower,Bathurst Quay,Island airport,Harbourf...             0.0   0.0625   
3                         Cabbagetown,St. James Town             0.0   0.0000   
4                                 Central Bay Street             0.0   0.0000   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0              0.0000        0.0000           0.000           0.0000   
1              0.0000        0.0000           0.000           0.0000   
2              0.0625        0.0625           0.125           0.1875   
3              0.0000        0.0000           0.000           0.0000   
4              0.0000        0.0000           0.000           0.0000   

   Airport Terminal  American Restaurant  Antique Shop  ...  Thai Restaurant  \
0             0.000             0.020833           0.0  ...         0.041667   
1             0.000             0.000000           0.0  ...         0.016949   
2             0.125             0.000000           0.0  ...         0.000000   
3             0.000             0.000000           0.0  ...         0.020833   
4             0.000             0.000000           0.0  ...         0.015873   

    Theater  Theme Restaurant  Trail  Train Station  \
0  0.010417               0.0    0.0            0.0   
1  0.000000               0.0    0.0            0.0   
2  0.000000               0.0    0.0            0.0   
3  0.000000               0.0    0.0            0.0   
4  0.000000               0.0    0.0            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                       0.010417               0.0                    0.0   
1                       0.016949               0.0                    0.0   
2                       0.000000               0.0                    0.0   
3                       0.000000               0.0                    0.0   
4                       0.015873               0.0                    0.0   

   Wine Bar  Yoga Studio  
0  0.000000     0.000000  
1  0.000000     0.000000  
2  0.000000     0.000000  
3  0.000000     0.000000  
4  0.015873     0.015873  

[5 rows x 207 columns]

In [179]:
dt_grouped.shape

(18, 207)

In [197]:
num_top_venues = 5

for hood in dt_grouped['Neighbourhood']:
    print("----"+hood+"----")
    temp = dt_grouped[dt_grouped['Neighbourhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
             venue  freq
0      Coffee Shop  0.10
1             Café  0.05
2       Restaurant  0.04
3  Thai Restaurant  0.04
4              Gym  0.03


----Berczy Park----
          venue  freq
0   Coffee Shop  0.08
1  Cocktail Bar  0.07
2        Bakery  0.05
3      Beer Bar  0.03
4    Restaurant  0.03


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0   Airport Service  0.19
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3   Harbor / Marina  0.06
4           Airport  0.06


----Cabbagetown,St. James Town----
                venue  freq
0         Coffee Shop  0.06
1  Italian Restaurant  0.04
2              Bakery  0.04
3         Pizza Place  0.04
4              Market  0.04


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.19
1      Sandwich Place  0.06
2                Café  0.05
3  Italian Restaurant  0.05
4        Burger Joint  0

In [198]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [199]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = dt_grouped['Neighbourhood']

for ind in np.arange(dt_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(dt_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2  CN Tower,Bathurst Quay,Island airport,Harbourf...       Airport Service   
3                         Cabbagetown,St. James Town           Coffee Shop   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Restaurant       Thai Restaurant   
1          Cocktail Bar                Bakery              Pharmacy   
2        Airport Lounge      Airport Terminal           Coffee Shop   
3                Market    Chinese Restaurant    Italian Restaurant   
4        Sandwich Place    Italian Restaurant                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0         Deli / Bodega        Clothing Store                   Gym   
1        Farmers Market           Cheese Shop            Restaurant   
2       Harbor / Marina         Boat or Ferry              Boutique   
3                Bakery                  Café                  Park   
4           Salad Place       Bubble Tea Shop          Burger Joint   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0            Steakhouse             Cosmetics Shop            Salad Place  
1    Seafood Restaurant                   Beer Bar           Concert Hall  
2   Rental Car Location           Sculpture Garden                Airport  
3            Restaurant                  Pet Store            Pizza Place  
4   Japanese Restaurant  Middle Eastern Restaurant      Indian Restaurant

In [200]:
# set number of clusters
kclusters = 4

dt_grouped_clustering = dt_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(dt_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 2, 0, 0, 0, 1, 0, 0, 0], dtype=int32)

In [201]:
# add clustering labels
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

dt_merged = dt_data

# merge manhattan_grouped with manhattan_data to add latitude/longitude for each neighborhood
dt_merged = dt_merged.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

dt_merged.head() # check the last columns!

Postal Code           Borough               Neighbourhood   Latitude  \
0         M4W  Downtown Toronto                    Rosedale  43.679563   
1         M4X  Downtown Toronto  Cabbagetown,St. James Town  43.667967   
2         M4Y  Downtown Toronto        Church and Wellesley  43.665860   
3         M5A  Downtown Toronto    Harbourfront,Regent Park  43.654260   
4         M5B  Downtown Toronto     Ryerson,Garden District  43.657162   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.377529               3                  Park            Playground   
1 -79.367675               0           Coffee Shop                Market   
2 -79.383160               0           Coffee Shop   Japanese Restaurant   
3 -79.360636               0           Coffee Shop                   Pub   
4 -79.378937               0        Clothing Store           Coffee Shop   

  3rd Most Common Venue      4th Most Common Venue  \
0                 Trail              Deli / Bodega   
1    Chinese Restaurant         Italian Restaurant   
2      Sushi Restaurant                 Restaurant   
3                Bakery                       Park   
4                  Café  Middle Eastern Restaurant   

         5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   
1                       Bakery                  Café                  Park   
2                      Gay Bar                   Pub                 Hotel   
3                         Café               Theater        Breakfast Spot   
4          Japanese Restaurant    Italian Restaurant        Cosmetics Shop   

  8th Most Common Venue     9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant                   Dog Run    Distribution Center  
1            Restaurant                 Pet Store            Pizza Place  
2           Yoga Studio  Mediterranean Restaurant            Men's Store  
3                 Hotel         French Restaurant      Health Food Store  
4       Bubble Tea Shop            Lingerie Store          Movie Theater

In [202]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(dt_merged['Latitude'], dt_merged['Longitude'], dt_merged['Neighbourhood'], dt_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

The observation that I can give from my analysis is that since Toronto has 103 neighborhoods when choosing a borough, our data set is much smaller, therefore when clustering is not as efficient, as we can observe. 